alf/algorithms/sarsa_algorithm.py

```diff
@@ -90,179 +210,343 @@ def __init__(self,
+        critics = []
+        new_states = []
+        if self._critic_bias is not None:
+            biases = self._critic_bias.get()
+        for i in range(self._num_replicas):
+            critic, state = critic_networks[i](inputs=inputs, state=states[i])
+            if self._critic_bias is not None:
+                critic += biases[i]
+            critics.append(critic)
+            new_states.append(state)
+        return critics, new_states
+
+    def rollout_step(self, time_step: TimeStep, state: SarsaState):
+        if self._on_policy:
+            return self._train_step(time_step, state)
+
+        if not self._is_rnn:
+            critic_states = state.critics
+        else:
+            _, critic_states = self._calc_critics(
+                self._critic_networks,
+                inputs=(state.prev_observation, time_step.prev_action),
+                states=state.critics)
+
+            not_first_step = time_step.step_type != StepType.FIRST
+
+            critic_states = common.reset_state_if_necessary(
+                state.critics, critic_states, not_first_step)
+
+        action_distribution, action, actor_state = self._get_action(
+            self._rollout_actor_network, time_step, state.actor)
+
+        if not self._is_rnn:
+            target_critic_states = state.target_critics
+        else:
+            _, target_critic_states = self._calc_critics(
+                self._target_critic_networks,
+                inputs=(time_step.observation, action),
+                states=state.target_critics)
 
-    def rollout(self, time_step: ActionTimeStep, state: SarsaState, mode):
-        not_first_step = tf.not_equal(time_step.step_type, StepType.FIRST)
-        prev_critic, critic_state = self._critic_network(
+        info = SarsaInfo(action_distribution=action_distribution)
+
+        rl_state = SarsaState(
+            prev_observation=time_step.observation,
+            prev_step_type=time_step.step_type,
+            actor=actor_state,
+            critics=critic_states,
+            target_critics=target_critic_states)
+
+        return AlgStep(action, rl_state, info)
+
+    def train_step(self, time_step: Experience, state: SarsaState):
+        return self._train_step(experience_to_time_step(time_step), state)
+
+    def _train_step(self, time_step: TimeStep, state: SarsaState):
+        not_first_step = time_step.step_type != StepType.FIRST
+        prev_critics, critic_states = self._calc_critics(
+            self._critic_networks,
             inputs=(state.prev_observation, time_step.prev_action),
-            step_type=state.prev_step_type,
-            network_state=state.critic)
-        critic_state = tf.nest.map_structure(
-            lambda new_s, s: tf.where(not_first_step, new_s, s), critic_state,
-            state.critic)
+            states=state.critics)
+
+        if self._critic_bias is not None:
+            self._critic_bias.update(
+                critics=[
+                    c - b
```

Comment:
Why -b here instead of just using c which is the critic?

Reply:
It is because b is already added to the original critic by `self._calc_critics()`